lib/controllers/adm/services.ts

```diff
@@ -233,11 +256,20 @@ export function GetServiceHandler(
  */
 export function GetService(serviceModel: ServiceModel): express.RequestHandler {
   const handler = GetServiceHandler(serviceModel);
+  const azureUserAttributesMiddleware = AzureUserAttributesMiddleware(
+    serviceModel
+  );
   const middlewaresWrap = withRequestMiddlewares(
     AzureApiAuthMiddleware(new Set([UserGroup.ApiServiceRead])),
+    ClientIpMiddleware,
+    azureUserAttributesMiddleware,
     requiredServiceIdMiddleware
   );
-  return wrapRequestHandler(middlewaresWrap(handler));
+  return wrapRequestHandler(
+    middlewaresWrap(
+      checkSourceIpForHandler(handler, (_, c, u, __) => ipTuple(c, u))
```

Comment:
the api here is a little bit ugly. that's because the middlewares cannot access other middlewares returned values so you had to wrap the whole handler. what about changin this:

> v3(request).then(r3)

to

> v3(request, r1, r2).then(r3)

in middleware.ts ? (so you can access previous middlewares result and get rid of the extractor function)

Reply:
how would you define the parameters accepted by a middleware then?

cannot be easily generalized since the parameters that gets passed to a middleware would depend on the order of previous middlewares

Reply:
ok. in the meatime I'm thinking about an alternative way to pass middlewares to handlers. ie. instead of having a fixed set of input parameter as with withRequestMiddlewares() we could curry the typed handler functions then pass one middleware at time until all parameters are satisfied (just brainstorming here).

Reply:
you should try to write down some example code on how that would work, I fear you'll lose type safety in a loop